From 0a37ff50579611230121d97ef9298ba3811bb573 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:35:43 -0400
Subject: [block] null_blk: fix use after free

Message-id: <1402670238-13916-56-git-send-email-snitzer@redhat.com>
Patchwork-id: 83807
O-Subject: [RHEL7.1 PATCH 055/150] block: null_blk: fix use after free
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit fc27691f3537a0df087214322467b642d1f6dedb
Author: Ming Lei <tom.leiming@gmail.com>
Date:   Thu May 1 15:12:36 2014 +0800

    block: null_blk: fix use after free

    entry(cmd->ll_list) may belong to new request once end_cmd()
    returns, so fix the bug with the patch.

    Without the change, it is easy to observe oops when
    doing null_blk(timer) test.

    Signed-off-by: Ming Lei <tom.leiming@gmail.com>
    Signed-off-by: Jens Axboe <axboe@fb.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/null_blk.c b/drivers/block/null_blk.c
index c9d1913..10d295a 100644
--- a/drivers/block/null_blk.c
+++ b/drivers/block/null_blk.c
@@ -203,8 +203,8 @@ static enum hrtimer_restart null_cmd_timer_expired(struct hrtimer *timer)
   entry = llist_reverse_order(entry);
   do {
    cmd = container_of(entry, struct nullb_cmd, ll_list);
-   end_cmd(cmd);
    entry = entry->next;
+   end_cmd(cmd);
   } while (entry);
  }
 
-- 
1.7.1